### PR TITLE
Another attempt to add an E2E test

### PR DESCRIPTION
**Pull request by bbrowning**
This is an attempt to push #121 to completion so we get some basic E2E testing for eventing-sources. It may take a couple of iterations here to get things passing.

### CODE REVIEW

**Comment by vaikas-google** [comment]
@n3wscott since you just cleaned up eventing e2e tests.

File: test/crd.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+				{
+					ConfigurationName: configName,
+					Percent:           100,
+				},
+			},
+		},
+	}
+}
+
+// Configuration returns a Configuration object in namespace with the name names.Config
```
Comment (vaikas-google) [inline test/crd.go]: nit. this comment is a bit confusing. names.Config doesn't seem to exist?

File: test/e2e/e2e.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+	err = CreateClusterRoleBinding(clients, crb, logger, cleaner)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (vaikas-google) [inline test/e2e/e2e.go]: Do we not want these cleaned up? As in, how come these are not added to the cleaner?

File: test/e2e/e2e.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		logs, err := PodLogs(clients, podName, containerName, logger)
+		if err != nil {
+			return true, err
+		}
+		return strings.Contains(string(logs), content), nil
+	})
+}
+
+// WaitForAllPodsRunning will wait until all pods in the given namespace are running
+func WaitForAllPodsRunning(clients *test.Clients, logger *logging.BaseLogger, namespace string) error {
```
Comment (vaikas-google) [inline test/e2e/e2e.go]: Does this block forever or does it timeout at some point? Or put another way, if something goes wrong, will things block forevs?

File: test/e2e/k8s_events_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	err = CreatePod(clients, test.NGinxPod(testNamespace), logger, cleaner)
+	if err != nil {
+		t.Fatalf("Failed to create Pod: %v", err)
+	}
+
+	WaitForAllPodsRunning(clients, logger, testNamespace)
+
+	err = WaitForLogContent(clients, logger, routeName, "user-container", "Created container")
+	if err != nil {
+		t.Fatalf("Events for container created not received: %v", err)
```
Comment (vaikas-google) [inline test/e2e/k8s_events_test.go]: bit of a nit, my understanding has been that errors should be logged with %s.

**Comment by n3wscott** [comment]
It looks like there is a lot of copy/pasta from the e2e test from eventing. I think we can let this in at the moment but I think we (knative) should build up a e2e framework that can be shared between the project. What do you think of that overall plan? 

So, let this in, but it is tech-debt.

File: test/e2e/k8s_events_test.go (hunk 4 above)
Comment (n3wscott) [reply test/e2e/k8s_events_test.go]: we have always used `%v` I think, when did this change?

**Comment by bbrowning** [comment]
There is definitely copy & paste here - both from the previous PR (121) and from Eventing repo in general. At the time, my goal was to just push the e2e work already started to a working state so that we'd get some e2e test coverage here.

I've resolved the merge conflicts here so it's unblocked from getting in. And I'll create an issue in knative/pkg to track pulling these common e2e test bits out. I believe other repos also have similar duplication of e2e helpers that would also benefit from consolidation.

**Comment by bbrowning** [comment]
@vaikas-google @n3wscott Note that the most recent unit tests run as part of this PR exposed the same data race observed on #191 - a lot of `DATA RACE` mentions at https://storage.googleapis.com/knative-prow/pr-logs/pull/knative_eventing-sources/178/pull-knative-eventing-sources-unit-tests/1093897152024809473/build-log.txt

**Comment by n3wscott** [comment]
minor conflicts^

**Comment by n3wscott** [comment]
which one wins? https://github.com/knative/eventing-sources/pull/121

File: test/cleanup.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (vaikas-google) [inline test/cleanup.go]: nit, 2019 in all the new files.

File: test/e2e/k8s_events_test.go (hunk 4 above)
Comment (bbrowning) [reply test/e2e/k8s_events_test.go]: It looks like `%v` is used all over the places in other knative repos when using `t.Fatalf`, so I'll leave this as-is for now.

File: test/e2e/e2e.go (hunk 2 above)
Comment (bbrowning) [reply test/e2e/e2e.go]: They are already added to the cleaner from the `CreateServiceAccount` and `CreateClusterRoleBinding` methods called here.

File: test/e2e/e2e.go (hunk 3 above)
Comment (bbrowning) [reply test/e2e/e2e.go]: This timeouts after 8 minutes - the `WaitForPodListState` method called here is in knative/pkg and the timeout is set there.

**Comment by bbrowning** [comment]
This initial e2e test is passing again and ready for an updated review. The expectation is that once we get an e2e suite bootstrapped others can start adding e2e tests for each source.

This is a full E2E test of the Kubernetes Events source wired up via Eventing. I know we've moved towards reducing test dependencies between repos in the other projects, but I left that dependency in for now as otherwise there's nowhere that confirms Eventing Sources and Eventing actually work together.

I believe the existing issue https://github.com/knative/pkg/issues/26 is probably the best place to track reducing this kind of copy/paste between e2e setups in each repo and I added a comment there as well as this comment that will link the two issues.

File: test/clients.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+	clients.Sources, err = sources.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return clients, nil
+}
+
+func buildClientConfig(kubeConfigPath string, clusterName string) (*rest.Config, error) {
```
Comment (srinivashegde86) [inline test/clients.go]: BuildclientConfig() can be used from pkg/test

File: test/cleanup.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+			}); err != nil {
+				c.logger.Errorf("Error: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// CleanupOnInterrupt will execute the function cleanup if an interrupt signal is caught
+func CleanupOnInterrupt(cleanup func(), logger *logging.BaseLogger) {
```
Comment (srinivashegde86) [inline test/cleanup.go]: CleanupOnInterrupt() can be used from pkg/test

File: test/e2e/e2e.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
+package e2e
```
Comment (srinivashegde86) [inline test/e2e/e2e.go]: missing file copyright header

File: test/e2e/e2e.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+	"github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	pkgTest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	servingV1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	rbacV1beta1 "k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	// Mysteriously required to support GCP auth (required by k8s libs).
```
Comment (srinivashegde86) [inline test/e2e/e2e.go]: I dont think we need this anymore

Comment (bbrowning) [reply test/e2e/e2e.go]: Serving and Eventing master still have this in as well. I can attempt to take it out and see what happens, but I assume something still needs it.

**Comment by bbrowning** [comment]
Thanks for the reviews!

**Pull request merged**